### PR TITLE
[#3131] Extend invalidating cache in protocol adapters.

### PR DESCRIPTION
**Pull request by n-deliyski**
Purge all cache entities from the ProtonBasedTenantClient on
database update operations at Tenant collection. It is enabled
via invalidate-cache-on-update flag at Tenant ext properties.

Signed-off-by: Nikolay Deliyski <Nikolay.Deliyski@bosch.io>

### CODE REVIEW

**Comment by kaniyan** [comment]
@n-deliyski Thanks for contributing and it looks good. There are few points to address:
1. In this PR, _invalidate-cache-on-update_ tenant level property is only being used to invalidate tenant elements. As per 2.i approach, it should also be extended to devices and credentials too. Could you please include that too?

2. Would you mind also updating hono documentation with information about _invalidate-cache-on-update_ property.

**Review by kaniyan — CHANGES_REQUESTED** [review]

**Comment by sophokles73** [comment]
> In this PR, invalidate-cache-on-update tenant level property is only being used to invalidate tenant elements. As per 2.i approach, it should also be extended to devices and credentials too. Could you please include that too?

My understanding from the discussion in the issue is that we do not need to purge devices and credentials from client caches as part of purging tenants. Or am I mistaken?

**Comment by kaniyan** [comment]
> > In this PR, invalidate-cache-on-update tenant level property is only being used to invalidate tenant elements. As per 2.i approach, it should also be extended to devices and credentials too. Could you please include that too?
> 
> My understanding from the discussion in the issue is that we do not need to purge devices and credentials from client caches as part of purging tenants. Or am I mistaken?

As per my understanding we need not purge devices and credentials from client caches while purging tenants during tenant update operation.

In the issue discussion I saw the below point.
>  Support the invalidate-cache-on-update flag in DeviceChangeNotification and CredentialsChangeNotification as well.
>    Use the invalidate-cache-on-update flag configured at the tenant level in the registry to control if other entities should be purged on changes, or


From the above my understanding is that the `invalidate-cache-on-update`at tenant level should also be used to set the _DeviceChangeNotification_ and _CredentialsChangeNotification_ during device and credentials update operation. I meant that based on this new flag in those change notifications the corresponding device and credentials cache elements should be removed from the corresponding caches.

Based on the @n-deliyski last comment on that issue, we always remove credential element during credentials update operation. In that case I would extend atleast this flag to _DeviceChangeNotification_ and not needed for _CredentialsChangeNotification_.

File: core/src/main/java/org/eclipse/hono/util/TenantConstants.java [hunk 1]
```diff
@@ -204,6 +204,11 @@ public final class TenantConstants extends RequestResponseApiConstants {
      */
     public static final String FIELD_EXT_MESSAGING_TYPE = "messaging-type";
 
+    /**
+     * The name of the property that indicates if cache invalidation is required on db update operations.
+     */
+    public static final String FIELD_EXT_INVALIDATE_CACHE_ON_UPDATE = "invalidate-cache-on-update";
+
```
[the remark this answers is not in the document]
Comment (n-deliyski) [reply core/src/main/java/org/eclipse/hono/util/TenantConstants.java:211]: done

File: clients/notification/src/main/java/org/eclipse/hono/notification/NotificationConstants.java [hunk 2]
```diff
@@ -57,6 +57,10 @@ public class NotificationConstants {
      * The field name of the JSON object that indicates if the entity is enabled.
      */
     public static final String JSON_FIELD_DATA_ENABLED = "enabled";
+    /**
+     * The field name of the JSON object that indicates if cache invalidation is required on db update operations.
+     */
+    public static final String JSON_FIELD_DATA_INVALIDATE_CACHE_ON_UPDATE = "invalidate-cache-on-update";
     /**
```
[the remark this answers is not in the document]
Comment (n-deliyski) [reply clients/notification/src/main/java/org/eclipse/hono/notification/NotificationConstants.java:64]: done

File: site/documentation/content/api/tenant/index.md [hunk 3]
```diff
@@ -316,6 +317,14 @@ The table below contains the properties that are used to configure a tenant's re
 | *mode*                   | *yes*     | *string*      | `-`           | The mode of the resource usage calculation. The default implementation supports two modes namely `days` and `monthly`. |
 | *no-of-days*             | *no*      | *number*      | `-`           | When the mode is set as `days`, then this value represents the length of an accounting period , i.e. the number of days over which the resource usage is to be limited. MUST be a positive integer.|
 
+### Extension properties Format
+
+The table below contains the extension properties that are used for addition configuration:
+
+| Name                          | Mandatory | JSON Type | Default Value | Description |
+|:------------------------------|:---------:|:----------|:--------------| :---------- |
+| *invalidate-cache-on-update*  |   *no*    | *boolean* | `false`       | Purge all cache entities from the ProtonBasedTenantClient on update operations at Tenant collection. |
```
Comment (kaniyan) [inline site/documentation/content/api/tenant/index.md:326]: How about a description that does not involve any underlying implementation details? May be something likes this: _When invalidate-cache-on-update is set to true, the clients must purge the cached tenant entity for update operations._

**Review by kaniyan — CHANGES_REQUESTED** [review]

File: site/documentation/content/api/tenant/index.md (hunk 3 above)
Comment (n-deliyski) [reply site/documentation/content/api/tenant/index.md:326]: I tried to followed the style of the other boolean properties and to give a bit more meaning. What about:
_If set to `true`, the clients must purge the cached tenant entity when they receive change notification for update operation._

Comment (kaniyan) [reply site/documentation/content/api/tenant/index.md:326]: @n-deliyski sounds good.

Comment (n-deliyski) [reply site/documentation/content/api/tenant/index.md]: done

File: site/documentation/content/api/tenant/index.md [hunk 4]
```diff
@@ -105,16 +105,17 @@ It is an error to include payload that is not of this type.
...
+| *trusted-ca*           | *no*      | *array*   | The list of trusted certificate authorities to use for validating certificates presented by devices of the tenant for authentication purposes. See [Trusted Certificate Authority Format]({{< relref "#trusted-ca-format" >}}) for a definition of the content model of the objects contained in the array. **NB** If the element is provided then the list MUST NOT be empty.                                                                                                                               |
+| *ext*                  | *no*      | *object*  | Arbitrary *extension* properties belonging to the tenant. See [Extension properties Format]({{< relref "#extension-properties-format" >}}) for details.                                                                                                                                                                                                                                                                                                                                                      |
```
Comment (kaniyan) [inline site/documentation/content/api/tenant/index.md:118]: Could you please revert the above formatting changes and make use of the existing formatting style.

**Review by kaniyan — CHANGES_REQUESTED** [review]

**Review by kaniyan — APPROVED** [review]
LGTM.
@n-deliyski Thank you for contributing :-)

**Pull request merged**